src/ssl.c

```diff
@@ -15468,24 +15565,22 @@ int wolfSSL_SetSession(WOLFSSL* ssl, WOLFSSL_SESSION* session)
             ret = WOLFSSL_FAILURE;
         }
         if (ret == WOLFSSL_SUCCESS) {
-            int error = 0;
             ref = session; /* keep copy of ref for later */
             session = (WOLFSSL_SESSION*)session->refPtr;
-            row = HashSession(ref->sessionID, ID_LEN, &error) % SESSION_ROWS;
-            if (error != 0) {
-                WOLFSSL_MSG("Hash session failed");
-                ret = WOLFSSL_FAILURE;
-            }
+            row = session->cacheRow;
         }
+        if (ret == WOLFSSL_FAILURE)
+            return WOLFSSL_FAILURE;
     }
 #endif
 
-    if (row < 0 || row >= SESSION_ROWS) {
+    if (row != INVALID_SESSION_ROW && (row < 0 || row >= SESSION_ROWS)) {
         return BAD_FUNC_ARG;
     }
 
     /* lock session cache row */
-    if (SESSION_ROW_LOCK(&SessionCache[row]) != 0) {
+    if (row != INVALID_SESSION_ROW &&
```

Comment:
This code was specifically designed to avoid XMALLOC/XFREE during the session lock. Please consider these changes.